### ISSUE TITLE
Nativewind styles not being applied intermittently

### ISSUE THREAD

**Issue by finallyblueskies** [nativewind/nativewind#924]
**Describe the bug**
Styles are not applied immediately when serving the app using Metro bundler. It usually takes a couple of refreshes/changes to relevant components before styles are applied.

**To Reproduce**
This is an intermittent bug and I'm not sure how to best recreate it. This is enough on my end:
1. Create a new Expo app with SDK 51
3. Follow nativewind expo router instructions to add nativewind to project
4. Add a component with nativewind styles
5. npx expo run:ios
6. Observe the rendered component in iOS simulator

**Expected behavior**
All styles should be applied.
- Checking node_modules nativewind cache confirms that the relevant CSS classes are generated.
- Using the nativewind installation verification helper returns no errors or warnings. 

Updating the components and/or refreshing a couple of times will suddenly pop the styling into existence.

**Screenshots**

https://github.com/marklawlor/nativewind/assets/14274086/d1607753-5c61-4ec8-bd6b-9d483255711b

**Additional context**
    "nativewind": "^4.0.1",
    "expo-router": "~3.5.14",
    "expo": "~51.0.8",
    "tailwindcss": "^3.4.3",

**Comment by coyksdev** [comment]
i have the this same issue

**Comment by tyrauber** [comment]
I am seeing the same issue with Nativewind 4.0.36.

 I thought it might be related to the use of clsx, but I am even seeing it with direct className assignment.  Style occaasstionally fail, whether in development, a release build, or pushing an OTA. 99.99% of the time, styles render, but occasionally I'll have a particularly stubborn class that will refuse to render.   In development, editing and saving the file, **temporarily** fixes the issue, but that is obviously not a solution for production. Release builds and OTA's are the most troubling because they are difficult to resolve. 

Here is an example I am wrestling with right now:

```
<View className="h-4 pt-4">
   <View pointerEvents="box-none" className="self-center h-[4px] w-8 bg-gray-300 dark:bg-gray-500" />
</View>
```

For whatever reason, this bit of code fails to render the background color, and was particularly difficult to fix.

* I do have a custom hex gray scale applied in my tailwind.config.js through extend colors gray, but I do not believe that is the issue. Even if I remove the custom colors and rebuild, the issue persists.

* I thought it might be the custom pixel height, but switching to h-2 does not resolve the issue on rebuild.

* In development, if I edit the file add important bang operators to the the background colors, `!bg-gray-300 dark:!bg-gray-500`, and save, the styles will render, but the issue will return with a new build.

* I thought it maybe specifying the hex bg inline, `bg-[#D2D1CE] dark:bg-[#95928D]`, would resolve the issue, but the issue persists on rebuild.

* Declaring the colors in my global.css as a custom class (even if I don't add it to the className on the component), will resolve the issue in development, but fail on rebuild.

What is most troubling, is that even if I can resolve the issue for a new build, the style may still fail to update on an OTA.

For this specific instance above, the only thing that has worked with 100% certainty for development, release and OTA,  is  restructuring the code to remove the pt-4 on the parent, add flex flex-row, and using border instead of bg:

```
<View className="flex flex-row items-center justify-center h-8">
  <View className="flex w-6 h-1 border-2 border-gray-300 dark:border-gray-500" />
</View>
```

I have absolutely loved working with tailwind css in react-native with Nativewind, but this bug has been downright painful. A lot of lost time debugging and rebuilding due to weird, occassional, stubborn rendering errors.

This is my experience and findings as I work through this issue. I would love confirmation and additional insight.

**Comment by storyicon** [comment]
same here

**Comment by trevorpfiz** [comment]
deleting the `.expo` folder and running `npx expo start --clear` seems to help

**Comment by tyrauber** [comment]
Isn't deleting the `.expo` folder the same as `expo start --clear`?

My issue/concern is not with Expo Go, but with the development-client (`expo run:ios -d`), production release builds (`eas build`) and OTA updates (`eas update`).   While deleting the .expo folder and rebuilding may temporarily resolve the issue, when cutting a release build, or an OTA, these stubborn issues revert.

**Comment by tyrauber** [comment]
@Darren120, Yes, although, this issue exists even without dynamic classnames as described above.  

But I have seen very similar issues (more frequently) when defining classNames dynamically with `clsx` or `tailwind-variants`, specifically when the tailwind className does not exist elsewhere in the app as defined in the tailwind.config `content` path.

**Comment by zarifabdalimov** [comment]
I came across a similar case when I was changing the colors in the tailwind.config. Steps to reproduce:

- Add color 
- Use color in App
- Works fine
- Change this color in tailwind.confg
- Check app -> nothing changes

It seems like `expo start --clear` solves this issue but I am not quite sure because it happened that I tried run it and it doesn't help

**Comment by 0xifis** [comment]
I have also noticed this bug happening exactly how @finallyblueskies and @tyrauber have described it. I can also confirm that I noticed the issue come up without dynamic classes. Also as reported before, it seems to creep into production builds sometimes. 

Context
- Expo SDK @ 51.0.8
- Tailwind 3.4.1
- Nativewind 4.0.36
- @babel/core 7.24.5

Also tried a bunch of either Nativewind v4 versions, Tailwind versions without much luck

**Comment by GrinZero** [comment]
same

**Comment by AyoCodess** [comment]
I has the same issue #883

**Comment by AyoCodess** [comment]
> related? https://github.com/marklawlor/nativewind/issues/883

Yes

**Comment by tyrauber** [comment]
Just ran into another one, with two different issues, that failed to resolve in an OTA - both a height and bg color:

```
export const Pagination = React.memo(() => (
  <View className="absolute  top-[120px] flex flex-row w-full justify-center items-center pb-2">
    {Array.from({ length: 3 }).map((_, index) => (
      <View key={index} className="mx-1 h-2 w-2 bg-gray-300 dark:bg-gray-500 rounded-full" />
    ))}
  </View>
));
```
The `top-[120px]` and `bg-gray-300 dark:bg-gray-500` worked in development, but failed to export and update on an OTA.

I was only able to resolve it by specifying the top position with a style prop and defining the colors in the global.css.

```
export const Pagination = React.memo(() => (
  <View style={{ top: 120 }} className="absolute flex flex-row w-full justify-center items-center pb-2">
    {Array.from({ length: 3 }).map((_, index) => (
      <View key={index} className="mx-1 h-2 w-2 tertiary-fill bg-gray-300 dark:bg-gray-500 rounded-full" />
    ))}
  </View>
));
```

**Comment by AyoCodess** [comment]
its super weird, when i changed some text in one of my components, by deleting some whitespace, my expo project still persevered the white space for that **1** instance across re-renders and builds. for other instances of the same text node, the white space is successfully deleted.

**Comment by av-erencelik** [comment]
I came across this issue with custom color classes, top classes and text align classes. Clearing the cache solves it but only for temporarily. Issue still happens after couple of edits.

**Comment by AyoCodess** [comment]
> I came across this issue with custom color classes, top classes and text align classes. Clearing the cache solves it but only for temporarily. Issue still happens after couple of edits.

Clearing the cache doesn't work for me unfortunately

**Comment by HarunKilic** [comment]
Did also come across this issue. Thought it was my environment setup or something that was causing it. 
As for my experience, production build has no issue, but results in unreliable results in development, as you can't really trust your changes cause of this issue.

**Comment by AyoCodess** [comment]
> results

I've ended up moving more and more to native styles, as I cant trust naivewind right now. The inconsistency is just risky to take on.

**Comment by HarunKilic** [comment]
> > results
> 
> I've ended up moving more and more to native styles, as I cant trust naivewind right now. The inconsistency is just risky to take on.

Did also end up moving to unistyles instead.

**Comment by aretrace** [comment]
I encounter this issue in an Expo project when running both the simulator and web instances simultaneously.

**Comment by DanSytnikov** [comment]
Same here

**Comment by Malberee** [comment]
same

**Comment by tyrauber** [comment]
Hey @marklawlor, sorry to ping you, but can you lend your expertise on this issue?  

We are seeing rare, but stubborn, style discrepancies between the development-client (expo run:ios -d), production release builds (eas build) and OTA updates (eas update). While deleting the .expo folder and rebuilding may temporarily resolve the issue, when cutting a release build, or an OTA, these stubborn issues revert.  More [details and examples above](https://github.com/nativewind/nativewind/issues/924#issuecomment-2122306893).  

Any insight you can provide would be greatly appreciated! :pray:

**Comment by alexwolf22** [comment]
Hi does anyone here have Nativewind for a preview or production EAS build working? If so what library versions are you using? I'm using these versions

```
"expo": "^51.0.0",
"nativewind": "2.0.11",
"react": "18.2.0",
"react-native": "0.74.2",
"tailwindcss": "3.3.2", (devDependencies)
"@babel/core": "^7.23.9",
```

For me Nativewind is working perfectly for an EAS development build, but does not work for a preview or production build. I need to figure out a resolution to this as soon as possible.

I've made issues about this on [Nativewind](https://github.com/nativewind/nativewind/issues/939) and [Expo](https://github.com/expo/eas-cli/issues/2429)

**Comment by tyrauber** [comment]
@alexwolf22, you are using the old version. You want to be using the carany release, [4.0.36](https://www.npmjs.com/package/nativewind/v/4.0.36).

**Comment by alexwolf22** [comment]
@tyrauber I tried using the new version of NativeWind but was not working on my development build.

Classnames styles are intermittently being applied. V4 does not seem to be reliable yet.

My issue is surrounding v2 not working on preview/production builds.

**Comment by alexwolf22** [comment]
I figure out my issue with V2 made comment here about it- https://github.com/nativewind/nativewind/issues/939#issuecomment-2173701012

Was not related to Nativewind source code.

**Comment by anhkieet** [comment]
I faced the same issue as you. Expo 51 and NativeWind 4.0.36

**Comment by chhornponleu** [comment]
I faced the same issue as you. Expo 51 and NativeWind 4.0.36

I noticed this, in app/index.tsx I have
```
<View className="flex flex-1 p-6 space-y-4">
      <View className="h-10 round rounded-lg items-center align-top" />
      <Button variant="primary">Hello</Button>
      <Button variant="secondary">Hello</Button>
 </View>
 ```
All built style inside Button component not applied until I add `<View className="h-10 round rounded-lg items-center align-top" />`. Only class names `h-10....align-top` got applied, the rest remain unapplied.

**Comment by pitops** [comment]
Have the exact same issue. Nativewind v4 (at least) seems very unreliable right now. Seems I will be moving away for a while.

**Comment by finallyblueskies** [comment]
@tyrauber have you had any luck fixing OTA updates? I've got an instance where styles in a component completely refuse to update with those, not even working intermittently.

**Comment by tyrauber** [comment]
@finallyblueskies, all I can suggest is change the layout and styles.  Finding a different approach to accomplish the design is the most effective way to resolve the specific issue in my experience.

**Comment by tyrauber** [comment]
@danstepanov, [super happy to see you have joined as a maintainer of Nativewind](https://github.com/nativewind/nativewind/issues/600#issuecomment-2192912430)! Also psyched to read that you `have a "stable" version of NativeWind v4 that we're testing with a small audience`.  

Does this new version address the intermittent style issues outlined above?  Specifically, occasional, rare issues related to different style generation in development, store releases and OTAs? 

If you want to push a new alpha pre-release, I'd be happy to test it out.

**Comment by diemer** [comment]
@pitops have you continued to have success, or has the issue cropped back up?

Can you share an example, even if it's pseudo code of what you mean regarding the tailwind.config paths?

**Comment by pitops** [comment]
@diemer i continued to have success really.

The main part is this in tailwind.config. This is crucial especially if using the expo app router where the root path is `.` instead of `src`

```
module.exports = {
  // NOTE: Update this to include the paths to all of your component files.
  content: ['./app/**/*.{js,jsx,ts,tsx}', './components/**/*.{js,jsx,ts,tsx}'],
```

**Comment by average-digital-developer** [comment]
I started using Nativewind yesterday. Got this bug inconsistently after installing new libs to the project - and only installing it, not changing a single line of code. Still can't find out what's going on.

**Comment by feliperugai** [comment]
I started using Nativewind yesterday. Got this bug inconsistently after installing new libs to the project - and only installing it, not changing a single line of code. Still can't find out what's going on.

**Comment by necmettindev** [comment]
Bump

**Comment by prince272** [comment]
Same problem here.

**Comment by rlesniak** [comment]
@danstepanov - any updates on this?

**Comment by danstepanov** [comment]
For now, run `npx expo start -c` to resolve this issue. 

I'm working with Mark on getting these tests passed for the version where we introduced a permanent solution to this problem.

Posting another State of NativeWind (Twitter) tomorrow with updates.

**Comment by rlesniak** [comment]
Thank you @danstepanov. Dev mode is less problematic - can be bypassed somehow. Worse with `eas build`. Do you know root cause of prod build issue? Is it the same reason as for dev mode?

**Comment by Morenoo1105** [comment]
Hey! Having the same issue over here.

On a development build everything looks nice and tidy, but when building for production, some of my padding styles went missing (mostly on TouchableOpacity componentes). Upon restarting the app, those styles came back now ScrollView component (and children) paddings disappeared!

The best solution seems to just move to native styles...

**Comment by tyrauber** [comment]
As @rlesniak stated, development is not really an issue. Clearing the cache and rebuilding resolves the issue in development. But `eas build` and `eas update` are impossible to resolve.

**Comment by danstepanov** [comment]
> Thank you @danstepanov. Dev mode is less problematic - can be bypassed somehow. Worse with `eas build`. Do you know root cause of prod build issue? Is it the same reason as for dev mode?

Haven't looked into this yet but can take a look today to try to gauge whether it stems from the same root cause.

If the symptoms are the same, it's highly likely this will be resolved with the upcoming minor version bump. I'll follow up once I gain a better understanding of the root issue though.

A link to a repro would be super helpful if not already provided above.

**Comment by necmettindev** [comment]
hi @danstepanov,

is there a release date for the new version?

**Comment by danstepanov** [comment]
> hi @danstepanov,
> 
> is there a release date for the new version?

There is not. However, I am shooting for an end of week release. I'll post on Twitter throughout the week re the progress of my work on the tests. As an FYI, Mark is the only one that can actually push a release.

https://x.com/nativewindcss

**Comment by felippewick** [comment]
I'm facing the same issue @tyrauber describes. Development is manageable but production issues with eas build and eas update remain. Only after a couple of restarts of the app the styles are applied.

**Comment by onyedikachi23** [comment]
> For now, run `npx expo start -c` to resolve this issue.
> 
> I'm working with Mark on getting these tests passed for the version where we introduced a permanent solution to this problem.
> 
> Posting another State of NativeWind (Twitter) tomorrow with updates.

Solution doesn't work in mine. I might have to switch to native styles instead.

**Comment by xinha-sh** [comment]
I am using `react-native-ide` and the issue still persists. `npx expo start -c` doesn't work for me

`nativewind: 4.1.1`

**Comment by lucas-garrido** [comment]
same issue, clearing doesn't help

**Comment by danstepanov** [comment]
We are closing all pre-v4.1 issues now that v4.1 is released. If the problem is still happening with the latest v4.1, [please open a new issue](https://github.com/nativewind/nativewind/blob/next/contributing.md#opening-an-issue) and reference this one in the description.

**Comment by Seanmclem** [comment]
I still get this, with 4.1, I get zero styles in my dev-client development build

**Comment by marklawlor** [comment]
@Seanmclem Please stop commenting on old closed issues. This issues were closed because the originally reported issue was resolved. You will not receive support this way and you are just annoying people with meaningless notifications.

If you have a issue you will need to create a new issue and follow the issue template

**Comment by onyedikachi23** [comment]
@marklawlor This issue was never resolved. You just closed it because of the no responses to your questions.

If it was resolved, people would get the solution provided, or do you want the same issue being opened repetitively?

**Comment by marklawlor** [comment]
@onyedikachi23 I would like the same issue opened repeatedly, this makes my life 100x easier.

People are very bad at identifying what the root cause is (that's not their job, it's mine) so I'd rather group the issues together rather than 10 people with different problems piling into the same issue

**Comment by bklingelfus** [comment]
I believe I have the same issue and I might have found one of the reasons why it happens. 

Context

- Expo SDK @ 52.0.17
- Tailwind 3.4.16
- Nativewind 4.1.23


The issue was happening in a very simple custom component for a styled text that essentially replaces the standard text element. The problem appeared inconsistently through out the app and even after clearing the cache or deleting the .expo it persisted. With the code below, I got the error with the text colors, but I tested with other styles such as margins and background colors and it also occurs with them. 

Here is the component:

```
const StyledText: React.FC<Props> = ({
	children,
	className,
	style,
	numberOfLines,
	ellipsizeMode,
}) => {
	// Render
	return (
		<Text
			numberOfLines={numberOfLines}
			ellipsizeMode={ellipsizeMode}
			style={[style]}
			className={`font-montserrat text-slate-950 ${className}`}
		>
			{children}
		</Text>
	);
};

export default StyledText;
```

So the error happens when you try to use a class that is already present (I have tested other colors and they work, but like I said it ain't predictable, sometimes a color might work right now but later it stops behaving as it used to). 

Usage error example:

```
<StyledText className="flex-grow font-montserrat-medium text-red-600">
	{someTextHere}
</StyledText>
```

The expected behaviour for Tailwind is that the class that comes later is the one that gets applied, but in my experience this is happening inconsistently with NativeWind. This also happens without custom components such as `className="text-black text-red-500"` in a Text element. 

Based on my experience and from the other comments, my guess is that for classes that require "on the fly" validation such as conflicting classes, dark mode ( dark:some-class ), dynamic values ( h-[some value] ), etc. the code is inconsistently  able to render the style.

My workaround is not ideal, but for my use case it worked since there aren't many places that I use these type of classes. Essentially I have to avoid it and the best way I have found is to have inner elements, which works a little bit better with Text elements since they can inherit parent Text styles.

Usage:

```
<StyledText className="flex-grow font-montserrat-medium">
	<Text className="text-red-600">{someTextHere}</Text>
</StyledText>
```

**Comment by onyedikachi23** [comment]
It's a pity NativeWind still marked this as closed even when the issue keeps gaining engagements.

It's funny that the bug is the style gets applied, and some time later, it won't be applied again. Such inconsistency in it's efficiency.

**Comment by marklawlor** [comment]
@onyedikachi23 Please create a new issue. I am unaware of any inconsistency issues, the issues people have reported have been caused by project specific issuse. Thats why I will not reopen this issue, because in nearly every-case its a project specific issue that needs a unique answer and should be addressed in its own issue.

**Comment by onyedikachi23** [comment]
I suspended a new personal project because of this. I guess I'd create a new issue when I continue working on it.

**Comment by kyawthura-gg** [comment]
If anyone still having this issue on v4, don't forget to add `nativewind/babel` presets in `babel.config.js` file. Here is how my `babel.config.js` file looks like:
```js
module.exports = function (api) {
  api.cache(true)
  return {
    presets: [
      [
        "babel-preset-expo",
        {
          jsxImportSource: "nativewind",
        },
      ],
      "nativewind/babel",
    ],
    plugins: ['react-native-reanimated/plugin'],
  }
}
```

**Comment by Ayowalexy** [comment]
The issue persists.
Works and, for whatever reason, stops working.

**Comment by FaresKrouma** [comment]
problem still exists.

**Comment by body20002** [comment]
I have the same issue.

**Comment by mohamedabdinasir** [comment]
For anyone still having this issue, you might be trying to add styles in a folder not in your content array in tailwind.config.js.
For example, if you are trying to style a button in the components folder, but the components folder is not in the array like this: ``` content: ['./app/**/*.{js,jsx,ts,tsx}', './components/**/*.{js,jsx,ts,tsx}'], ```, then your code is likely not to update on the fly.



```
/** @type {import('tailwindcss').Config} */
module.exports = {
    content: ['./app/**/*.{js,jsx,ts,tsx}', './components/**/*.{js,jsx,ts,tsx}'],
    presets: [require('nativewind/preset')],
    theme: {
        extend: {
            colors: {
                'brand-red': '#DE0724',
                'charcoal-gray': '#333333',
            },
        },
    },
    plugins: [],
}

```
Check out this issue as well: https://github.com/nativewind/nativewind/issues/1428

@FaresKrouma @AyoCodess @Ayowalexy @onyedikachi23

**Comment by danstepanov** [comment]
Neither this nor #1364 have repros so I'm closing and locking both. Both of these issues seem to have to do with misconfigured projects. The installation instructions are now up to date and both `npx rn-new --nativewind` and `npx rn-new --expo-router --nativewind` so likely not an issue with NativeWind. If you create a minimal repro and it doesn't work, please feel free to open a new issue.
